SimPEG/regularization/base.py

```diff
@@ -151,10 +290,8 @@ def deriv(self, m):
             R(m) = \\mathbf{W^\\top W (m-m_\\text{ref})}
 
         """
-
-        mD = self.mapping.deriv(self._delta_m(m))
-        r = self.W * (self.mapping * (self._delta_m(m)))
-        return mD.T * (self.W.T * r)
+        r = self.W * self.f_m(m)
+        return self.f_m_deriv(m).T * (self.W.T * r)
 
     @utils.timeIt
     def deriv2(self, m, v=None):
```

Comment:
Just a thought, but I wonder if we could implement the regularizations as a `HasModel` class, so we can make use of the automated clearing of certain items if the model is updated, (i.e. here we can avoid creating the derivative mapping matrix at each call to `deriv2`.